src/backend/distributed/clock/causal_clock.c

```diff
@@ -0,0 +1,639 @@
+	ereport(DEBUG1, (errmsg("node(%u) transaction clock %lu:%u",
+							PostPortNumber, localClockValue->logical,
+							localClockValue->counter)));
+
+	/* Check if there are any remote clock values */
+	if (globalClockValue)
+	{
+		globalClockValue = LargerClock(globalClockValue, localClockValue);
+	}
+	else
+	{
+		globalClockValue = localClockValue;
+	}
+
+	ereport(DEBUG1, (errmsg("final global transaction clock %lu:%u",
+							globalClockValue->logical,
+							globalClockValue->counter)));
+	return globalClockValue;
+}
+
+
+/*
+ * AdjustClocksToTransactionHighest Sets the clock value of all the nodes, participated
+ * in the PREPARE of the transaction, to the highest clock value of all the nodes.
+ */
+static void
+AdjustClocksToTransactionHighest(List *nodeConnectionList, ClusterClock
+								 *transactionClockValue)
+{
+	StringInfo queryToSend = makeStringInfo();
+
+	/* Set the clock value on participating worker nodes */
+	appendStringInfo(queryToSend,
+					 "SELECT pg_catalog.citus_internal_adjust_local_clock_to_remote"
+					 "('(%lu, %u)'::pg_catalog.cluster_clock);",
+					 transactionClockValue->logical, transactionClockValue->counter);
+
+	ExecuteRemoteCommandInConnectionList(nodeConnectionList, queryToSend->data);
+	AdjustLocalClock(transactionClockValue);
+}
+
+
+/*
+ * PrepareAndSetTransactionClock polls all the transaction-nodes for their respective clocks,
+ * picks the highest clock and returns it via UDF citus_get_transaction_clock. All the nodes
+ * will now move to this newly negotiated clock.
+ */
+static ClusterClock *
+PrepareAndSetTransactionClock(void)
+{
+	if (!EnableClusterClock)
+	{
+		/* citus.enable_cluster_clock is false */
+		ereport(NOTICE, (errmsg("GUC enable_cluster_clock is off")));
+		return NULL;
+	}
+
+	RequireTransactionBlock(true, "citus_get_transaction_clock");
+
+	dlist_iter iter;
+	List *transactionNodeList = NIL;
+	List *nodeList = NIL;
+
+	/* Prepare the connection list */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		WorkerNode *workerNode = FindWorkerNode(connection->hostname, connection->port);
+
+		/* Skip the node if we already in the list */
+		if (list_member_int(nodeList, workerNode->groupId))
+		{
+			continue;
+		}
+
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		Assert(transaction->transactionState != REMOTE_TRANS_NOT_STARTED);
+
+		/* Skip a transaction that failed */
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, add to the list, otherwise we skip
+		 * prepare since the connection didn't perform any write (read-only)
+		 */
+		if (ConnectionModifiedPlacement(connection))
+		{
+			nodeList = lappend_int(nodeList, workerNode->groupId);
+			transactionNodeList = lappend(transactionNodeList, connection);
+		}
```

Comment:
I feel this check is incorrect. Even if no modification has been this node, eg. due to only reading from this node, it shows up in the connection list, is still part of the causal ordering for other modifications made in this transaction.

eg. when you read a value from a distributed table and deduce the value you want to store based on what you have read should show causality between the changes made prior to your query on the first node, which could be arbitrarily ahead of behind on the clock.

Reply:
Agree, that's why I included the transaction-coordinator without checking its properties. Bottom line is all participants are causally related.  Removing this check.